Changes for eval2

Hyperion101010 opened the pull request:
Seems like the bugs are now taken care of, I would like to merge my changes in the official repo.

Commit:
Add diagnostic for code compilation.

This commit adds code diagnostics for testing of
Solidity files during development. This checks are
triggered on 'save' and 'open' actions in the IDE.
Also contains tests for validating the same.

Signed-off-by: Shivam Balikondwar <shivambalikondwar@icloud.com>

Commit:
Add documentation for diagnostics.

Added documentation on the diagnostics feature
and some gifs for easy illustrations.
README is also updated for the same.

Signed-off-by: Shivam Balikondwar <shivambalikondwar@icloud.com>

Commit:
Bug Fixes in Diagnostics implementation.

Fixed the clear message issue of diagnostic messages.
Some typos and links in the documentation.

Signed-off-by: Shivam Balikondwar <shivambalikondwar@icloud.com>

seanyoung commented on .github/workflows/rust.yaml:
```diff
@@ -0,0 +1,20 @@
+          override: true
+      - run: cargo build
+        working-directory: solang-server
+      - run: cargo fmt --all
```
I think it's a good idea to have "cargo clippy" here as well.

Here you have cargo build *before* cargo fmt, which is a good idea. I think I'll steal that for the solang repo!

Hyperion101010 replied:
Sure, ha ha alright!

Hyperion101010 replied:
done.

seanyoung commented on solang-server/src/ServerUtils/backend.rs:
```diff
@@ -0,0 +1,286 @@
+#[warn(dead_code)]
+impl Backend {
+    pub fn search_line(data: &str, val: usize) -> (usize, usize) {
+        let mut line = Vec::new();
```
- This function does not need to be `pub`. 
- The name is not descriptive. How about `file_offset_to_line_column` (just a suggestion)
- Function needs doccomment explaining what it does
- The idea of building of building the `line` vector and then doing binary search is that it is faster when doing multiple scans. Here you build the `line` vector for each call, which is actually less efficient than a straight scan of the file.

I think it should either cache the `line` vector somewhere or just scan the data until offset for the right number of lines.

Hyperion101010 replied:
done.

seanyoung commented on solang-server/src/ServerUtils/backend.rs:
```diff
@@ -0,0 +1,286 @@
+        (count, col - 1)
+    }
+
+    pub fn convert_diagnos(ns: ast::Namespace, filecache: &mut FileCache) -> Vec<Diagnostic> {
```
- Not sure this function needs to be public.
- Abbreviating diagnositic to diagnos is a bit confusing
- Needs a doccomment explaining what it does

Every function in this file could do with doccomment explaining what it does

Hyperion101010 replied:
done

seanyoung commented on solang-server/src/main.rs:
```diff
@@ -0,0 +1,18 @@
+use tower_lsp::{LspService, Server};
+
+#[allow(non_snake_case)]
+mod ServerUtils;
```
Can we rename `ServerUtils` into something rust accepts without `#[allow(non_snake_case)]` please

Hyperion101010 replied:
done

seanyoung reviewed (changes requested):
Looks great, some minor changes please.

Now that `tower-lsp` directory is no longer used, please also remove it.

Commit:
Changes for review.

 Now using linear search for line, coloumn conversion.
 Added some proper comments at respective places.
 Fixes for clippy, snake-case, pub declarations.
 Added debug for diagnostics, some typos and readme fix.

Signed-off-by: Shivam Balikondwar <shivambalikondwar@icloud.com>

seanyoung commented on solang-server/src/serverutils/backend.rs:
```diff
@@ -0,0 +1,279 @@
+                ast::Level::Info => DiagnosticSeverity::Information,
+                ast::Level::Warning => DiagnosticSeverity::Warning,
+                ast::Level::Error => DiagnosticSeverity::Error,
+                ast::Level::Debug => DiagnosticSeverity::Hint,
```
debug is not a hint, debug diagnostics should be ignored.

Hyperion101010 replied:
done

seanyoung reviewed (changes requested):
tower-lsp is still present
PR cannot be merged (see github message)
